### PR TITLE
XCVM-283 - Part 1

### PR DESCRIPTION
**Pull request by cor**
Splitting this up into multiple PRs as I want to keep it reviewable and prevent huge rebases

### CODE REVIEW

File: code/parachain/frame/cosmwasm/src/lib.rs [hunk 1]
```diff
@@ -498,1084 +513,836 @@ pub mod pallet {
...
+	let mut shared = Pallet::<T>::do_create_vm_shared(gas, InitialStorageMutability::ReadOnly);
+	let query_request = serde_json::from_slice(&query_request)
+		.map_err(|e| CosmwasmVMError::Rpc(format!("{}", e)))?;
+	Pallet::<T>::sub_level_dispatch(
+		&mut shared,
+		contract.clone(),
+		contract,
+		Default::default(),
+		|vm| {
+			cosmwasm_system_query(vm, query_request)?
```
Comment (hussein-aitlahcen) [inline code/parachain/frame/cosmwasm/src/lib.rs:537]: I think this is invalid and we must also hook it right?

File: code/parachain/frame/cosmwasm/src/lib.rs [hunk 2]
```diff
@@ -498,1084 +513,836 @@ pub mod pallet {
...
-						verify_items.push((signature, message, public_key)),
-					_ => return false,
-				}
-			}
+		// Else, the contract is not a pallet. We continue with the normal wasmi vm creation
+		// process.
+		let info = Self::contract_info(&contract)?;
+
+		// Reinstrument the code if needed
+		let code = Self::do_check_for_reinstrumentation(info.code_id)?.map_or(
```
Comment (hussein-aitlahcen) [inline code/parachain/frame/cosmwasm/src/lib.rs:1021]: Why are we not doing all that in the cache miss branch just like the previous code? We now check for instrumentation and load from `InstrumentedCode` even if the code is already present in the cache?

File: code/parachain/frame/cosmwasm/src/lib.rs [hunk 3]
```diff
@@ -498,1084 +513,836 @@ pub mod pallet {
...
+					.ok_or(Error::<T>::CodeNotFound)?;
+				(contract_info, code_info)
+			},
+		};
+		let ibc_port = if code_info.ibc_capable {
+			Some(Pallet::<T>::do_compute_ibc_contract_port(address))
+		} else {
+			None
+		};
+		let pinned = vm.shared.cache.code.contains_key(&contract_info.code_id);
```
Comment (hussein-aitlahcen) [inline code/parachain/frame/cosmwasm/src/lib.rs:1307]: We can return pinned from the above match and have a constant `true` for pallets.

**Review by hussein-aitlahcen — CHANGES_REQUESTED** [review]

File: code/parachain/frame/cosmwasm/src/lib.rs (hunk 2 above)
Comment (cor) [reply code/parachain/frame/cosmwasm/src/lib.rs:1021]: Ah yes you are correct. Forgot to fix the order. Will fix

File: code/parachain/frame/cosmwasm/src/lib.rs (hunk 1 above)
Comment (aeryz) [reply code/parachain/frame/cosmwasm/src/lib.rs:537]: yes, because this currently doesn't support native contracts.

**Review by aeryz — APPROVED** [review]

File: code/parachain/frame/cosmwasm/src/crypto.rs [hunk 4]
```diff
@@ -0,0 +1,132 @@
+use crate::{Config, Pallet, SUBSTRATE_ECDSA_SIGNATURE_LEN};
+
+// TODO(cor): move these out of the `impl` as they do not refer to `self` or `Self`.
+use sp_core::{ecdsa, ed25519};
+impl<T: Config> Pallet<T> {
+	pub(crate) fn do_secp256k1_recover_pubkey(
```
[the remark this answers is not in the document]
Comment (cor) [reply code/parachain/frame/cosmwasm/src/crypto.rs:6]: Yes I plan on removing most `do_` prefixes in part 2

File: code/parachain/frame/cosmwasm/src/lib.rs (hunk 1 above)
Comment (cor) [reply code/parachain/frame/cosmwasm/src/lib.rs:537]: Yes but this is unaltered by this PR

File: code/parachain/frame/cosmwasm/src/lib.rs (hunk 2 above)
Comment (cor) [reply code/parachain/frame/cosmwasm/src/lib.rs]: Fixed in d896f3567e1556f537608470dfa293682d0e71c0

**Review by hussein-aitlahcen — APPROVED** [review]

**Review by aeryz — APPROVED** [review]
This is for you mergify!

**Review by hussein-aitlahcen — APPROVED** [review]

**Comment by benluelo** [comment]
@mergify refresh

**Comment by cor** [comment]
@mergify refresh

**Review by PoisonPhang — APPROVED** [review]

**Comment by cor** [comment]
@mergify refresh

**Review by aeryz — APPROVED** [review]

**Review by aeryz — APPROVED** [review]

**Review by hussein-aitlahcen — APPROVED** [review]

**Review by aeryz — APPROVED** [review]

**Review by benluelo — APPROVED** [review]

**Review by PoisonPhang — APPROVED** [review]

**Pull request merged**